torch/csrc/jit/passes/loop_unrolling.cpp

```diff
@@ -0,0 +1,217 @@
+  for (int64_t i = 0; i < limit; ++i, ++it) {
+    for (Block *subblock : it->blocks()) {
+      i += limitedBlockSize(subblock, limit);
+    }
+    if (it == end) {
+      return i;
+    }
+  }
+  return limit;
+}
+
+bool isSmallBlock(Block *body) {
+  return limitedBlockSize(body, kMaxBodySize + 1) <= kMaxBodySize;
+}
+
+at::optional<int64_t> getConstantLength(Node *loop) {
+  Value *trip_count = loop->inputs().at(0);
+  if (trip_count->node()->kind() != prim::Constant)
+    return at::nullopt;
+  return {trip_count->node()->t(attr::value).toCLong()};
+}
+
+// XXX: This function can only be called with a loop that is guaranteed to execute EXACTLY ONCE.
+void inlineBody(Node *loop) {
+  auto graph = loop->owningGraph();
+  auto body = loop->blocks().at(0);
+  WithInsertPoint insert_point_guard { loop };
+
+  std::unordered_map<Value*, Value*> value_map;
+  auto get_value = [&](Value *v) {
+    auto it = value_map.find(v);
+    if (it != value_map.end())
+      return it->second;
+    return v;
+  };
+
+  // Loop node has extra (max_iters, initial_cond) inputs,
+  // body has an extra (loop_counter) input.
+  for (size_t i = 2; i < loop->inputs().size(); ++i) {
+    value_map[body->inputs()[i - 1]] = loop->inputs()[i];
+  }
+
+  for (Node *orig : body->nodes()) {
+    Node *clone = graph->insertNode(graph->createClone(orig, get_value));
+    for (size_t i = 0; i < orig->outputs().size(); ++i) {
+      value_map[orig->outputs()[i]] = clone->outputs()[i];
+    }
+  }
+  for (size_t i = 0; i < loop->outputs().size(); ++i) {
+    loop->outputs().at(i)->replaceAllUsesWith(get_value(body->outputs().at(i + 1)));
+  }
+  // XXX: it is extremely important to destroy the loop in here. DCE might not be able
+  // to conclude that it's safe, because the loop might contain side effects.
+  loop->destroy();
+}
+
+void repeatBody(Block *body, int64_t times) {
+  // We will be adding nodes to the body, so cache the initial start and end.
+  // XXX: they are both inclusive
+  auto body_start = body->nodes().begin();
+  auto body_end = std::prev(body->nodes().end());
+  auto graph = body->owningGraph();
+  WithInsertPoint insert_point_guard { body->return_node() };
+
+  std::unordered_map<Value*, Value*> value_map;
+  auto get_value = [&](Value *v) {
+    auto it = value_map.find(v);
+    if (it != value_map.end())
+      return it->second;
+    return v;
+  };
+
+  for (int64_t i = 1; i < times; ++i) {
+    // Update loop-carried values
+    // NB: note that we don't need to worry about the loop counter, because we've
+    //     replaced it with a loop-carried variable
+    JIT_ASSERT(body->inputs().size() == body->outputs().size());
+    for (size_t i = 1; i < body->inputs().size(); ++i) {
+      value_map[body->inputs()[i]] = get_value(body->outputs()[i]);
+    }
+
+    // Clone the nodes
+    for (auto it = body_start; it != std::next(body_end); ++it) {
+      Node *orig = *it;
+      Node *clone = graph->insertNode(graph->createClone(orig, get_value));
```

Comment:
Maybe we should add `createClone(Node*, std::unordred_map<Value*, Value*> & map)` that both does map lookup for inputs and map insertion for new values, since this pattern pops up in many places.

Reply:
+1 good idea